src/platform_impl/macos/mod.rs

```diff
@@ -1 +1,526 @@
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Menu {
+    pub ns_menu: id,
+}
+
+impl Menu {
+    pub fn new() -> Self {
+        let ns_menu: *mut Object;
+        unsafe {
+            ns_menu = NSMenu::alloc(nil).autorelease();
+            let () = msg_send![ns_menu, setAutoenablesItems: NO];
+        }
+        Self { ns_menu }
+    }
+
+    pub fn append(&self, item: &dyn crate::MenuEntry) {
+        self.add_menu_item(item, AddOp::Append)
+    }
+
+    pub fn prepend(&self, item: &dyn crate::MenuEntry) {
+        self.add_menu_item(item, AddOp::Insert(0))
+    }
+
+    pub fn insert(&self, item: &dyn crate::MenuEntry, position: usize) {
+        self.add_menu_item(item, AddOp::Insert(position))
+    }
+
+    fn add_menu_item(&self, item: &dyn crate::MenuEntry, op: AddOp) {
+        let ns_menu_item = match item.type_() {
+            MenuEntryType::Submenu => {
+                let submenu = item.as_any().downcast_ref::<crate::Submenu>().unwrap();
+                submenu.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Normal => {
+                let menuitem = item.as_any().downcast_ref::<crate::MenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Check => {
+                let menuitem = item.as_any().downcast_ref::<crate::CheckMenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Predefined => {
+                let menuitem = item.as_any().downcast_ref::<crate::PredefinedMenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+        };
+
+        unsafe {
+            match op {
+                AddOp::Append => self.ns_menu.addItem_(ns_menu_item),
+                AddOp::Insert(position) => {
+                    msg_send![self.ns_menu, insertItem: ns_menu_item atIndex: position as NSInteger]
+                }
+            }
+        }
+    }
+
+    pub fn remove(&self, item: &dyn crate::MenuEntry) {
+        let ns_menu_item = match item.type_() {
+            MenuEntryType::Submenu => {
+                let submenu = item.as_any().downcast_ref::<crate::Submenu>().unwrap();
+                submenu.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Normal => {
+                let menuitem = item.as_any().downcast_ref::<crate::MenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Check => {
+                let menuitem = item.as_any().downcast_ref::<crate::CheckMenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+            MenuEntryType::Predefined => {
+                let menuitem = item.as_any().downcast_ref::<crate::PredefinedMenuItem>().unwrap();
+                menuitem.0.ns_item_for_menu(self)
+            }
+        };
+        unsafe {
+            let () = msg_send![self.ns_menu, removeItem: ns_menu_item];
+        }
```

Comment:
`Menu` should also contain a `children` field and you should remove from it after this call